fdbserver/MockGlobalState.actor.cpp

```diff
@@ -246,8 +316,160 @@ Future<Void> MockStorageServer::run() {
+	usedDiskSpace -= totalByteSize;
+	clearRangeTotalBytes(range, beginShardBytes, endShardBytes);
+	return totalByteSize;
+}
+
+void MockStorageServer::get(KeyRef const& key, int64_t bytes) {
+	// If the read yields no value, randomly sample the empty read.
+	int64_t bytesReadPerKSecond = std::max(bytes, SERVER_KNOBS->EMPTY_READ_PENALTY);
+	metrics.notifyBytesReadPerKSecond(key, bytesReadPerKSecond);
+}
+
+int64_t MockStorageServer::getRange(KeyRangeRef const& range, int64_t beginShardBytes, int64_t endShardBytes) {
+	int64_t totalByteSize = estimateRangeTotalBytes(range, beginShardBytes, endShardBytes);
+	// For performance concerns, the cost of a range read is billed to the start key and end key of the
+	// range.
+	if (totalByteSize > 0) {
+		int64_t bytesReadPerKSecond = std::max(totalByteSize, SERVER_KNOBS->EMPTY_READ_PENALTY) / 2;
+		metrics.notifyBytesReadPerKSecond(range.begin, bytesReadPerKSecond);
+		metrics.notifyBytesReadPerKSecond(range.end, bytesReadPerKSecond);
+	}
+	return totalByteSize;
+}
+
+int64_t MockStorageServer::estimateRangeTotalBytes(KeyRangeRef const& range,
+                                                   int64_t beginShardBytes,
+                                                   int64_t endShardBytes) {
+	int64_t totalByteSize = 0;
+	auto ranges = serverKeys.intersectingRanges(range);
+
+	// use the beginShardBytes as partial size
+	if (ranges.begin().begin() < range.begin) {
+		ranges.pop_front();
+		totalByteSize += beginShardBytes;
+	}
+	// use the endShardBytes as partial size
+	if (ranges.end().begin() < range.end) {
+		totalByteSize += endShardBytes;
+	}
+	for (auto it = ranges.begin(); it != ranges.end(); ++it) {
+		totalByteSize += it->cvalue().shardSize;
+	}
+	return totalByteSize;
+}
+
+void MockStorageServer::clearRangeTotalBytes(KeyRangeRef const& range, int64_t beginShardBytes, int64_t endShardBytes) {
+	auto ranges = serverKeys.intersectingRanges(range);
+
+	// use the beginShardBytes as partial size
+	if (ranges.begin().begin() < range.begin) {
+		auto delta = std::min(ranges.begin().value().shardSize, (uint64_t)beginShardBytes);
+		ranges.begin().value().shardSize -= delta;
+		ranges.pop_front();
+	}
+	// use the endShardBytes as partial size
+	if (ranges.end().begin() < range.end) {
+		auto delta = std::min(ranges.end().value().shardSize, (uint64_t)endShardBytes);
+		ranges.end().value().shardSize -= delta;
+	}
+	for (auto it = ranges.begin(); it != ranges.end(); ++it) {
+		it->value().shardSize = 0;
+	}
+}
+
+void MockStorageServer::notifyWriteMetrics(KeyRef const& key, int64_t size) {
+	// update write bandwidth and iops as mock the cost of writing a mutation
+	StorageMetrics s;
+	// FIXME: remove the / 2 and double the related knobs.
+	s.writeBytesPerKSecond = mvccStorageBytes(size) / 2;
+	s.iosPerKSecond = 1;
+	metrics.notify(key, s);
+}
+
+void MockStorageServer::signalFetchKeys(const KeyRangeRef& range, int64_t rangeTotalBytes) {
+	fetchKeysRequests.send({ range, rangeTotalBytes });
```

Comment:
I think it isn't necessary to use a `PromiseStream` for this. Instead, this method can call:
```
if (!self->allShardStatusEqual(params.keys, MockShardStatus::COMPLETED)) {
    actors.add(waitFetchKeysFinish(this, params));
}
```
Then the `serveMockStorageServer` actor is unnecessary.